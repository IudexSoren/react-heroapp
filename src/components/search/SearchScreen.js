import { useMemo, useState } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from "../../hooks/useForm";
import HeroCard from "../heroes/HeroCard";
import { getHeroesByName } from '../../selectors/getHeroesByName';


const SearchScreen = ({ history }) => {

  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    searchText: q
  });

  const { searchText } = formValues;
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    //Establecer par√°metro en el URL
    history.push(`?q=${ searchText }`);
  }

  return (
    <div>
      <h1>Search a hero</h1>
      <hr/>
      <div className="row">
        <div className="col-5">
          <form onSubmit={ handleSearch }>
            <input
              type="text"
              autoComplete="off"
              placeholder="Find a hero"
              className="form-control"
              name="searchText"
              value={ searchText }
              onChange={ handleInputChange }
            />
            <button
              className="btn btn-outline-success btn-block mt-3 w-100"
              type="submit"
            >Search</button>
          </form>
        </div>
        <div className="col-7 d-flex justify-content-center flex-wrap">
          <h4 className="text-center w-100">Results</h4>
          <hr className="w-100"/>
          { (q === '') &&
            <div className="alert alert-info w-75">
              Search a hero
            </div>
          }
          { (q !== '' && heroesFiltered.length === 0) &&
            <div className="alert alert-warning w-75">
              No results for { q }
            </div>
          }
          {
            heroesFiltered.map((hero) =>
              <HeroCard key={ hero.id } { ...hero } />
            )
          }
        </div>
      </div>
    </div>
  )
}

export default SearchScreen
