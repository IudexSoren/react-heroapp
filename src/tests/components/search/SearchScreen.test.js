import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

describe('Pruebas de SearchScreen', () => {

  const history = {
    push: jest.fn(),
    replace: jest.fn()
  }

  test('Debe renderizar correctamente el componente', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search'] }>
        <Route
          exact
          path='/search'
          component={ () => <SearchScreen history={ history } /> }
        />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text()).toBe('Search a hero');
  });

  test('Debe mostrar un héroe y el valor del input debe basarse en el argumento de la búsqueda', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search?q=batman'] }>
        <Route
          exact
          path='/search'
          component={ () => <SearchScreen history={ history } /> }
        />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper.find('HeroCard').exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe mostrar el mensaje de error al no encontrar un héroe con los argumentos especificados en el URL', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search?q=chico%20percebe'] }>
        <Route
          exact
          path='/search'
          component={ () => <SearchScreen history={ history } /> }
        />
      </MemoryRouter>
    );

    expect(wrapper.find('.alert-warning').text()).toBe('No results for chico percebe');
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe ejecutar el PUSH del history', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={ ['/search'] }>
        <Route
          exact
          path='/search'
          component={ () => <SearchScreen history={ history } /> }
        />
      </MemoryRouter>
    );
    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'spider'
      }
    });
    wrapper.find('form').simulate('submit', {
      preventDefault(){}
    });

    expect(history.push).toHaveBeenCalledWith('?q=spider');
  });

});
