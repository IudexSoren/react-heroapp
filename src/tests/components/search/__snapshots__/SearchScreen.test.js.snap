// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de SearchScreen Debe mostrar el mensaje de error al no encontrar un héroe con los argumentos especificados en el URL 1`] = `
<div>
  <h1>
    Search a hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a hero"
          type="text"
          value="chico percebe"
        />
        <button
          className="btn btn-outline-success btn-block mt-3 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 d-flex justify-content-center flex-wrap"
    >
      <h4
        className="text-center w-100"
      >
        Results
      </h4>
      <hr
        className="w-100"
      />
      <div
        className="alert alert-warning w-75"
      >
        No results for 
        chico percebe
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas de SearchScreen Debe mostrar un héroe y el valor del input debe basarse en el argumento de la búsqueda 1`] = `
<div>
  <h1>
    Search a hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-success btn-block mt-3 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 d-flex justify-content-center flex-wrap"
    >
      <h4
        className="text-center w-100"
      >
        Results
      </h4>
      <hr
        className="w-100"
      />
      <div
        className="card mx-1 my-1 text-white bg-dark"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
            </div>
            <a
              className="link-light"
              href="/hero/dc-batman"
              onClick={[Function]}
            >
              Más...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas de SearchScreen Debe renderizar correctamente el componente 1`] = `
<div>
  <h1>
    Search a hero
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-success btn-block mt-3 w-100"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7 d-flex justify-content-center flex-wrap"
    >
      <h4
        className="text-center w-100"
      >
        Results
      </h4>
      <hr
        className="w-100"
      />
      <div
        className="alert alert-info w-75"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
